### PR TITLE
Add stats formatting for riak-admin and http stats output

### PR DESCRIPTION
**Pull request by bowrocker**
YZ stats are not currently displayed on the riak http stats endpoint, or by riak-admin status. Add YZ stats formatting appropriate for console output with 'riak-admin status' and the htts /stats endpoint. The stats namespace for Search is transformed to /search/[query | index]/... instead of 'yokozuna'.

This PR also depends on changes to riak_kv in order to publish the stats to the 2 interfaces:
https://github.com/basho/riak_kv/pull/839

Example stats output:

search_stat_ts : 1392310824
search_query_fail_count : 0
search_query_fail_one : 0
search_index_throughput_count : 0
search_index_throughput_one : 0
search_index_fail_count : 0
search_index_fail_one : 0
search_query_throughput_count : 0
search_query_throughput_one : 0
search_query_latency_min : 0.0
search_query_latency_max : 0.0
search_query_latency_median : 0.0
search_query_latency_percentile_95 : 0.0
search_query_latency_percentile_99 : 0.0
search_index_latency_min : 0.0
search_index_latency_max : 0.0
search_index_latency_median : 0.0
search_index_latency_percentile_95 : 0.0
search_index_latency_percentile_99 : 0.0

### CODE REVIEW

**Comment by rzezeski** [comment]
- [ ] code review
- [x] compile/eunit
- [x] dialyzer
- [ ] riak test
- [ ] manual check of HTTP stats
- [ ] manual check of `riak-admin status`

File: src/yz_stat.erl [hunk 1]
```diff
@@ -58,6 +58,17 @@ get_stats() ->
         Error -> Error
     end.
 
+%% @doc Return formatted stats
+-spec get_formatted_stats() -> [term()].
+get_formatted_stats() ->
+    case yokozuna:is_enabled(index) andalso ?YZ_ENABLED of
```
Comment (rzezeski) [inline src/yz_stat.erl]: No need for the `yokozuna:is_enabled(index)` check. That check is to allow users to just disabled indexing or querying without disabling all of Yokozuna. We will want to be able to retrieve stats even if index is disabled.

File: src/yz_stat.erl [hunk 2]
```diff
@@ -175,3 +186,65 @@ update(StatUpdate) ->
...
     ok.
+
+%% @private
+%%
+%% @doc Format stats for legacy stats blob
+-spec format_stats([term()]) -> [term()].
+format_stats(Stats) ->
+    lists:flatten([format(Stat) || Stat <- Stats]).
+
+format({{_, search, fail},[{count, Count}, {one, One}]}) ->
```
Comment (rzezeski) [inline src/yz_stat.erl]: Add a type spec for this function and mark as private.

File: src/yz_stat.erl [hunk 3]
```diff
@@ -175,3 +186,65 @@ update(StatUpdate) ->
...
+     {search_index_throughput_one, One}];
+format({{_, search, latency}, ValueList}) ->
+    [format_query_latency(V) || V <- ValueList];
+format({{_, index, latency}, ValueList}) ->
+    [format_index_latency(V) || V <- ValueList];
+format({yokozuna_stat_ts, TS}) ->
+    [{search_stat_ts, TS}];
+format(T) -> T.
+
+format_query_latency({min, Value}) ->
```
Comment (rzezeski) [inline src/yz_stat.erl]: Add type specs for all these functions and mark as private. You don't have to spec them to death. E.g. `{atom, term()}` is probably okay for `format_query_latency`.

**Comment by rzezeski** [comment]
I think we should rename the internal stat names to query as well to be consistent with external.

Also, looking at Riak KV it seems it stores the formatted stats in the cache. In fact, KV considers the formatted stats to be "legacy". I think this was because at one point @russelldb was working on moving to the new system but we still needed to present the stats in the old way. Thus `riak_kv_stat_bc` was created to map the legacy stat names to the new stat namespace. `riak_kv_stat_bc:produce_stats` is what is registered with the stat cache. Perhaps we should also cache the formatted stats to be consistent. It looks like we may also be able to utilize the `riak_kv_stat_bc:bc_stat` code since it exports all functions. This would remove the need for all the manual pattern matching and replace it with a list like that found in `riak_kv_stat_bc:legacy_stats_map` (I wouldn't call it legacy though in this case, maybe just `stats_map`).   This would avoid repeating code especially as we add more stats.

File: src/yz_stat.erl [hunk 4]
```diff
@@ -30,7 +30,7 @@
 %% -type microseconds() :: integer().
 -define(SERVER, ?MODULE).
 -define(NOTIFY(A, B, Type, Arg),
-        folsom_metrics:notify_existing_metric({?YZ_APP_NAME, A, B}, Arg, Type)).
+        folsom_metrics:notify_existing_metric({search, A, B}, Arg, Type)).
```
Comment (rzezeski) [inline src/yz_stat.erl]: I'm pretty sure all other systems using stats use the application name as first part of namespace. Part of me likes having the root of the namespace be `search` as that's what yokozuna is for but a bigger part of me wants to stay consistent with other areas of Riak which all namespace based on application.

Change it back to `?YZ_APP_NAME`. The renaming will happen in the `stats_map`.

File: src/yz_stat.erl [hunk 5]
```diff
@@ -51,8 +51,14 @@ register_stats() ->
     ok.
 
 %% @doc Return current aggregation of all stats.
--spec get_stats() -> proplists:proplist() | {error, term()}.
+-spec get_stats() -> proplists:proplist() | [].
```
Comment (rzezeski) [inline src/yz_stat.erl]: `proplists:proplist()` should cover the empty list.

File: src/yz_stat.erl [hunk 6]
```diff
@@ -85,6 +91,41 @@ search_fail() ->
 search_end(ElapsedTime) ->
     update({search_end, ElapsedTime}).
 
+%% @doc Optionally produce stats map based on ?YZ_ENABLED
+%%
```
Comment (rzezeski) [inline src/yz_stat.erl]: Remove empty comment line.

File: src/yz_stat.erl [hunk 7]
```diff
@@ -85,6 +91,41 @@ search_fail() ->
...
 
+%% @doc Optionally produce stats map based on ?YZ_ENABLED
+%% 
+-spec stats_map() -> [] | proplists:proplist().
+stats_map() ->
+     stats_map(?YZ_ENABLED).
+
+%% @doc Map to format stats for legacy "blob" if YZ_ENABLED,
+%% else []
+%%
```
Comment (rzezeski) [inline src/yz_stat.erl]: Empty comment line.

File: src/yz_stat.erl [hunk 8]
```diff
@@ -154,15 +195,15 @@ stats() ->
...
+     {['query', fail], spiral},
+     {['query', latency], histogram},
+     {['query', throughput], spiral}
     ].
 
 %% @private
 -spec stat_name(riak_core_stat_q:path()) -> riak_core_stat_q:stat_name().
 stat_name(Name) ->
-    list_to_tuple([?YZ_APP_NAME|Name]).
+    list_to_tuple([search|Name]).
```
Comment (rzezeski) [inline src/yz_stat.erl]: Revert back to `?YZ_APP_NAME`.

**Comment by rzezeski** [comment]
The `yz_stats` test is failing. Let me know if you have trouble
running the riak tests. This particular PR is a bit tricker than just
running `verify.sh` because you need a custom riak_kv. Also, you'll
want to merge develop into this branch as it's too far behind to work.

The `riak-admin` status works:

```
# ~/yz-verify/rt/riak_yz/dev/dev2/bin/riak-admin status | grep search
riak_search_version : <<"1.4.1-22-ga3893df">>
search_query_throughput_count : 0
search_query_throughput_one : 0
search_query_latency_min : 0
search_query_latency_max : 0
search_query_latency_median : 0
search_query_latency_95 : 0
search_query_latency_99 : 0
search_query_latency_999 : 0
search_index_throughput_count : 0
search_index_throughtput_one : 0
search_index_fail_count : 0
search_index_fail_one : 0
search_index_latency_min : 0
search_index_latency_max : 0
search_index_latency_median : 0
search_index_latency_95 : 0
search_index_latency_99 : 0
search_index_latency_999 : 0
```

But there was an error for HTTP stats.

```
# curl 'http://localhost:10018/stats'
<html><head><title>500 Internal Server Error</title></head><body><h1>Internal Server Error</h1>The server encountered an error while processing this request:<br><pre>{error,
    {error,function_clause,
        [{mochijson2,json_encode_string,
             [{yokozuna,query,throughput},{encoder,null,false}],
             [{file,"src/mochijson2.erl"},{line,173}]},
         {mochijson2,'-json_encode_proplist/2-fun-0-',3,
             [{file,"src/mochijson2.erl"},{line,166}]},
         {lists,foldl,3,[{file,"lists.erl"},{line,1248}]},
         {mochijson2,json_encode_proplist,2,
             [{file,"src/mochijson2.erl"},{line,170}]},
         {riak_kv_wm_stats,produce_body,2,
             [{file,"src/riak_kv_wm_stats.erl"},{line,76}]},
         {webmachine_resource,resource_call,3,
             [{file,"src/webmachine_resource.erl"},{line,186}]},
         {webmachine_resource,do,3,
             [{file,"src/webmachine_resource.erl"},{line,142}]},
         {webmachine_decision_core,resource_call,1,
             [{file,"src/webmachine_decision_core.erl"},{line,48}]}]}}</pre><P><HR><ADDRESS>mochiweb+webmachine web server</ADDRESS></body></html>
```

**Comment by rzezeski** [comment]
Something weird is going on here. The commits all occur twice with different hashes. Perhaps a bad rebase?

**Comment by bowrocker** [comment]
New PR:

https://github.com/basho/yokozuna/pull/318

**Pull request closed without merging**